### PR TITLE
#825 - Fix plugin reload on agent

### PR DESCRIPTION
**Pull request by mdaliejaz**

### CODE REVIEW

**Comment by mdaliejaz** [comment]
Personal is [green](http://go02.thoughtworks.com:8153/go/pipelines/value_stream_map/personal-ali/24).

File: plugin-infra/go-plugin-infra/src/com/thoughtworks/go/plugin/infra/listeners/PluginsZipUpdater.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    @Override
+    public void pluginJarUpdated(PluginFileDetails pluginFileDetails) {
+        pluginsZip.create();
+    }
+
+    @Override
+    public void pluginJarRemoved(PluginFileDetails pluginFileDetails) {
+        pluginsZip.create();
+    }
+}
```
Comment (jyotisingh) [inline plugin-infra/go-plugin-infra/src/com/thoughtworks/go/plugin/infra/listeners/PluginsZipUpdater.java]: The implementation worries me a bit. Reason being, now for every new plugin that gets added/removed/updated the whole jar would be recreated everytime, ie. if 10 new plugins get added, the plugin jar would be created 10 times. This won't be pretty for someone using a lot of plugins.

Comment (mdaliejaz) [inline plugin-infra/go-plugin-infra/src/com/thoughtworks/go/plugin/infra/listeners/PluginsZipUpdater.java]: Agreed.
Will change the implementation.

File: plugin-infra/go-plugin-infra/src/com/thoughtworks/go/plugin/infra/commons/PluginsZip.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            return computeMd5DigestOfPlugins();
+        }
+        return md5DigestOfPlugins;
+    }
+
+    private String computeMd5DigestOfPlugins() {
+        try {
+            String digestOfBundledFolder = md5DigestOfFolderContent(bundledPlugins);
+            String digestOfExternalFolder = md5DigestOfFolderContent(externalPlugins);
+            File tempFileToStoreMd5 = File.createTempFile("tempFileToStoreMd5", ".txt");
```
Comment (srinivasupadhya) [inline plugin-infra/go-plugin-infra/src/com/thoughtworks/go/plugin/infra/commons/PluginsZip.java]: do you need to create a new file. can we create md5 by string concatenation.

File: plugin-infra/go-plugin-infra/src/com/thoughtworks/go/plugin/infra/monitor/DefaultPluginJarLocationMonitor.java [hunk 3]
```diff
@@ -181,40 +183,50 @@ public void run() {
 
         private Set<PluginFileDetails> loadAndNotifyPluginsFrom(File pluginDirectory, Set<PluginFileDetails> knownPluginFiles, boolean isBundledPluginsLocation) {
             Set<PluginFileDetails> currentPluginFiles = getDetailsOfCurrentPluginFilesFrom(pluginDirectory, isBundledPluginsLocation);
-            notifyListenersOfRemovedPlugins(currentPluginFiles, knownPluginFiles);
-            notifyListenersOfUpdatedPlugins(currentPluginFiles, knownPluginFiles);
-            notifyListenersOfAddedPlugins(currentPluginFiles, knownPluginFiles);
+            boolean notifyListenersOfRemovedPlugins = notifyListenersOfRemovedPlugins(currentPluginFiles, knownPluginFiles);
+            boolean notifyListenersOfUpdatedPlugins = notifyListenersOfUpdatedPlugins(currentPluginFiles, knownPluginFiles);
+            boolean notifyListenersOfAddedPlugins = notifyListenersOfAddedPlugins(currentPluginFiles, knownPluginFiles);
+            if (notifyListenersOfRemovedPlugins || notifyListenersOfUpdatedPlugins || notifyListenersOfAddedPlugins) {
```
Comment (srinivasupadhya) [inline plugin-infra/go-plugin-infra/src/com/thoughtworks/go/plugin/infra/monitor/DefaultPluginJarLocationMonitor.java]: can this be made more simpler? computing 3 booleans & checking looks a little clumpsy.

File: plugin-infra/go-plugin-infra/test/com/thoughtworks/go/plugin/infra/DefaultPluginManagerTest.java [hunk 4]
```diff
@@ -416,6 +419,17 @@ public void shouldThrowExceptionIfMatchingExtensionVersionNotFound() throws Exce
...
 
+    @Test
+    public void shouldAddPluginChangeListener() throws Exception {
+        DefaultPluginManager pluginManager = new DefaultPluginManager(monitor, registry, mock(GoPluginOSGiFramework.class), jarChangeListener, applicationAccessor, pluginWriter, pluginValidator, systemEnvironment, pluginsZipListener);
+        pluginManager.startInfrastructure();
+
+        InOrder inOrder = inOrder(monitor);
+
+        inOrder.verify(monitor).addPluginJarChangeListener(jarChangeListener);
+//        inOrder.verify(monitor).addPluginJarChangeListener(pluginsZipUpdater);
```
Comment (srinivasupadhya) [inline plugin-infra/go-plugin-infra/test/com/thoughtworks/go/plugin/infra/DefaultPluginManagerTest.java]: intentional? if yes delete if you make another commit.

File: plugin-infra/go-plugin-infra/src/com/thoughtworks/go/plugin/infra/monitor/DefaultPluginJarLocationMonitor.java (hunk 3 above)
Comment (mdaliejaz) [inline plugin-infra/go-plugin-infra/src/com/thoughtworks/go/plugin/infra/monitor/DefaultPluginJarLocationMonitor.java]: No.
While it looks like a simple check similar to *currentPluginFiles!=knownPluginFiles* will work, it won't!

We need the three boolean values from the three methods since the current and known plugin files might have same name but different md5 (plugin update case). These methods take care of the md5 changes while comparing the plugin files and it would be duplication and unnecessary work to do that again over here. I thought it better to check on the boolean than do the check again.

File: plugin-infra/go-plugin-infra/test/com/thoughtworks/go/plugin/infra/DefaultPluginManagerTest.java (hunk 4 above)
Comment (mdaliejaz) [inline plugin-infra/go-plugin-infra/test/com/thoughtworks/go/plugin/infra/DefaultPluginManagerTest.java]: Oops. Yeah. The behaviour changed. That line is no longer required :)
Will delete that.

File: plugin-infra/go-plugin-infra/src/com/thoughtworks/go/plugin/infra/commons/PluginsZip.java (hunk 2 above)
Comment (mdaliejaz) [inline plugin-infra/go-plugin-infra/src/com/thoughtworks/go/plugin/infra/commons/PluginsZip.java]: No. Yes, we can.
